crates/evm/core/src/utils.rs

```diff
@@ -99,6 +108,118 @@ pub fn gas_used(spec: SpecId, spent: u64, refunded: u64) -> u64 {
+            };
+            if !ctx.external.should_use_create2_factory(&mut ctx.evm, &mut inputs) {
+                return old_handle(ctx, inputs);
+            }
+
+            // Sanity check that CREATE2 deployer exists.
+            let code_hash = ctx.evm.load_account(DEFAULT_CREATE2_DEPLOYER)?.0.info.code_hash;
+            if code_hash == KECCAK_EMPTY {
+                return Ok(FrameOrResult::Result(FrameResult::Create(CreateOutcome {
+                    result: InterpreterResult {
+                        result: InstructionResult::Revert,
+                        output: "missing CREATE2 deployer".into(),
+                        gas: Gas::new(inputs.gas_limit),
+                    },
+                    address: None,
+                })))
+            }
+
+            // Generate call inputs for CREATE2 factory.
+            let mut call_inputs = get_create2_factory_call_inputs(salt, *inputs);
+
+            // Call inspector to change input or return outcome.
+            if let Some(outcome) = ctx.external.call(&mut ctx.evm, &mut call_inputs) {
+                create2_overrides_inner
+                    .borrow_mut()
+                    .push((ctx.evm.journaled_state.depth(), call_inputs.clone()));
+                return Ok(FrameOrResult::Result(FrameResult::Call(outcome)));
+            }
+
+            // Push data about current override to the stack.
+            create2_overrides_inner
+                .borrow_mut()
+                .push((ctx.evm.journaled_state.depth(), call_inputs.clone()));
+
+            let mut frame_or_result = ctx.evm.make_call_frame(&call_inputs);
+
+            if let Ok(FrameOrResult::Frame(frame)) = &mut frame_or_result {
+                ctx.external
+                    .initialize_interp(&mut frame.frame_data_mut().interpreter, &mut ctx.evm)
+            }
+            frame_or_result
+        });
+
+    let create2_overrides_inner = create2_overrides.clone();
+    let old_handle = handler.execution.insert_call_outcome.clone();
+
+    handler.execution.insert_call_outcome =
+        Arc::new(move |ctx, frame, shared_memory, mut outcome| {
+            // If we are on the depth of the latest override, handle the outcome.
+            if create2_overrides_inner
+                .borrow()
+                .last()
+                .map_or(false, |(depth, _)| *depth == ctx.evm.journaled_state.depth())
+            {
+                let (_, call_inputs) = create2_overrides_inner.borrow_mut().pop().unwrap();
+                outcome = ctx.external.call_end(&mut ctx.evm, &call_inputs, outcome);
+
+                // Decode address from output.
+                let address = match outcome.instruction_result() {
+                    return_ok!() => {
+                        // SAFETY: we assume here that successful CREATE2 factory call will always
+                        // return a valid address.
+                        Some(Address::try_from(outcome.output().as_ref()).unwrap())
```

Comment:
Shouldn't we return an error / EVM revert instead of assuming this?

Reply:
yeah, makes sense, updated